Add support for per-prefix transmission throttling

Meorawr opened the pull request:
In patch 4.4.0 and 10.2.7 Blizzard have tightened the restrictions on addon comms to add a per-prefix throttle across all chat types, effectively restricting them to one message per second with a small accrued burst capacity.

The SendAddonMessage APIs now return an enum result code which includes information if this client-side throttle has been applied to a submitted message. With it, we can now properly handle throttling in CTL and avoid situations where addon messages would be dropped for exceeding it.

This PR takes into consideration the discussion on Discord and takes a slightly different approach to the other open one by instead implementing the concept of a "blocked" pipe.

A pipe enters the "blocked" state whenever a message at its head is attempted to be sent off, and a throttle result code is returned from the API.

When transitioning to this state, the pipe is removed from the transmission ring of its parent priority and is instead placed into a separate (and new) blocked ring. This prevents the despool logic from seeing blocked pipes and pointlessly attempting to re-send on them.

Periodically - currently every 0.35s - the contents of the blocked rings in each priority are reintegrated back into the transmission rings, allowing us to attempt re-transmission of queued messages.

This means there may be some added latency when a prefix entered a blocked state if the API was about to perhaps unblock it, but this also allows us to reallocate bandwidth that would be consumed by priorities that are fully blocked to others that can more readily use it. The value of 0.35s was chosen almost arbitrarily and could be tuned later if found to be a bit high.

It's important to also note that we specifically don't consider the 'ChannelThrottle' return code a retryable error condition. The reasoning here is that this throttle isn't new, and the API appears to enqueue these messages for submission anyway but with a hefty delay by grouping them into batches. There are additional server-side throttles (and a confirmed bug) that mean these sometimes may not make it through to their destination however, and so for consistency with the underlying API we treat it as a failure condition.

Aside from prefix throttling, there's a few other small changes.

- Failure to send a message either due to an error or throttling no longer consumes bandwidth that had been allocated to a priority.

- Priorities that enter a blocked or empty state now release their bandwidth back to the global pool for redistribution immediately, instead of waiting until there's no data queued up whatsoever. This is required to deal with edge cases involving priorities sending many small messages on one prefix infinitely accumulating bandwidth.

- Transmission logic has been centralized into a new PerformSend function to minimize the number of call sites individually needing to remember to toggle boolean variables with each Send call.

- Queued transmissions no longer apply checks to see if the player is in a group or raid. The API has dedicated return codes for this condition and has been tested to not trigger erroneous system message spam if attempting to send a message to either chat type while not being in a group. This is not the case for guilds, however the library never checked this case previously so one hasn't been added.

- User-supplied callbacks are now supplied an accurate 'didSend' parameter that will be false if the API returns a non-throttle-related error code.

- User-supplied callbacks are additionally now supplied the new result code as a third parameter. For Classic Era, we synthesize one from a subset of the enum values based off the boolean result that the API will still be providing there for now.

- User-supplied callbacks no longer let errors blow things up in an uncontrolled manner by being subject to securecall wrapping. This is also consistently applied irrespective of whether or not the send itself was immediate or queued.

- Some compatibility with the pre-8.0 global SendAddonMessage API was removed as it's no longer needed.

Meorawr committed:
Add support for per-prefix transmission throttling

In patch 4.4.0 and 10.2.7 Blizzard have tightened the restrictions on
addon comms to add a per-prefix throttle across all chat types,
effectively restricting them to one message per second with a small
accrued burst capacity.

The SendAddonMessage APIs now return an enum result code which includes
information if this client-side throttle has been applied to a submitted
message. With it, we can now properly handle throttling in CTL and
avoid situations where addon messages would be dropped for exceeding it.

This PR takes into consideration the discussion on Discord and takes a
slightly different approach to the other open one by instead
implementing the concept of a "blocked" pipe.

A pipe enters the "blocked" state whenever a message at its head is
attempted to be sent off, and a throttle result code is returned from
the API.

When transitioning to this state, the pipe is removed from the
transmission ring of its parent priority and is instead placed into a
separate (and new) blocked ring. This prevents the despool logic from
seeing blocked pipes and pointlessly attempting to re-send on them.

Periodically - currently every 0.35s - the contents of the blocked
rings in each priority are reintegrated back into the transmission
rings, allowing us to attempt re-transmission of queued messages.

This means there may be some added latency when a prefix entered a
blocked state if the API was about to perhaps unblock it, but this
also allows us to reallocate bandwidth that would be consumed by
priorities that are fully blocked to others that can more readily
use it. The value of 0.35s was chosen almost arbitrarily and could
be tuned later if found to be a bit high.

It's important to also note that we specifically don't consider the
'ChannelThrottle' return code a retryable error condition. The reasoning
here is that this throttle isn't new, and the API can return this value
but still sometimes send off the message albeit subject to additional
server-side throttling checks - and also at the time of writing, a
server-side bug that causes it to more aggressively throttle than
it actually should.

Aside from prefix throttling, there's a few other small changes.

- Failure to send a message either due to an error or throttling no
  longer consumes bandwidth that had been allocated to a priority.

- Priorities that enter a blocked or empty state now release their
  bandwidth back to the global pool for redistribution immediately,
  instead of waiting until there's no data queued up whatsoever. This
  is required to deal with edge cases involving priorities sending
  many small messages on one prefix infinitely accumulating bandwidth.

- Transmission logic has been centralized into a new PerformSend
  function to minimize the number of call sites individually needing
  to remember to toggle boolean variables with each Send call.

- Queued transmissions no longer apply checks to see if the player
  is in a group or raid. The API has dedicated return codes for this
  condition and has been tested to not trigger erroneous system
  message spam if attempting to send a message to either chat type
  while not being in a group. This is not the case for guilds, however
  the library never checked this case previously so one hasn't been
  added.

- User-supplied callbacks are now supplied an accurate 'didSend'
  parameter that will be false if the API returns a non-throttle-related
  error code.

- User-supplied callbacks are additionally now supplied the new result
  code as a third parameter. For Classic Era, we synthesize one from a
  subset of the enum values based off the boolean result that the API
  will still be providing there for now.

- User-supplied callbacks no longer let errors blow things up in an
  uncontrolled manner by being subject to securecall wrapping. This
  is also consistently applied irrespective of whether or not the
  send itself was immediate or queued.

- Some compatibility with the pre-8.0 global SendAddonMessage API was
  removed as it's no longer needed.

Nevcairiel commented on AceComm-3.0/ChatThrottleLib.lua:
```diff
@@ -353,6 +423,15 @@ function ChatThrottleLib.OnUpdate(this,delay)
 	end
 
+	-- Integrate blocked queues back into their rings periodically.
+	if self.BlockedQueuesDelay >= 1 then
```
I had been thinking about this delay last night, and wondered if using less then 1s would be better to lower latency in case the window expires a lot sooner - however with channel comms re-throttling on failed attempts, maybe thats a bad idea?

Meorawr replied:
Options are either lower the timeout or remove it entirely.

As I'd noted in the PR - because blocked pipes are appended to the _end_ of the priority, we have an effect where unblocked pipes implicitly have priority over blocked ones.

This means if the timer were removed and there's other traffic in a priority bracket, it's reasonably likely that we won't try re-sending data from the blocked pipes too quickly anyway, as first the round-robin despool logic is going to want to visit the unblocked pipes toward the start of the ring before getting to the blocked ones.

As such, the only circumstances at which we might unnecessarily try to re-send data too quickly is if we find ourselves in the situation where a priority bracket has very few unblocked pipes.

Also worth noting that the despooling itself occurs once every 0.08s. We could bump that to about 0.2s, which means that if the blocked pipe timer were removed that we're only going to potentially re-send on blocked pipes at most 5 times before the prefix is guaranteed to unthrottle itself.

I suspect increasing the latency of the despool by 120ms wouldn't really be too noticeable given that addon comms themselves are going to be subject to variable transmission latency anyway, and that any transmissions through CTL are already fully dependent upon its own view of available bandwidth.

Meorawr replied:
Blocked queues delay lowered to 0.35s in the squashed version.

Rationale for this figure is that;

1. We want to allow priorities to sit in a blocked state and relinquish their bandwidth for use by other prios that have actively-sendable data. If we removed the timer, this wouldn't be possible.
2. The original 1s value is quite a long time to wait between retries and in some cases may add a lot of artificial delay in scenarios where a prefix peaks over its allocated message count but and then regains a message slot on the next frame or so.
3. Too low values may result in increased CPU load. Once blocked queues are integrated back into the prio ring, the OnUpdate loop potentially has to walk all of them again in cases where the prio _almost exclusively_ consists of queues that are still in the blocked state when the API returns from its Send call. The chosen 0.35s figure should mitigate this, as we expect that within 3 attempts a prefix will unblock itself.

Nevcairiel commented on AceComm-3.0/ChatThrottleLib.lua:
```diff
@@ -292,38 +321,78 @@ end
-			Prio.avail = Prio.avail - msg.nSize
 			bMyTraffic = true
-			msg.f(unpack(msg, 1, msg.n))
+			sendResult = MapToSendResult(select(-1, true, msg.f(unpack(msg, 1, msg.n))))
```
I was wondering if it made sense to put this into a function of its own, something like this:

result = PerformSend(msg.f, unpack(msg, 1, msg.n))

That way the bMyTraffic stuff can also go in that function, and the return value shuffling is in one place and not repeated, in case we maybe can clean that up at some magical future date.

Its not a required change, but might make it a bit cleaner.

Meorawr committed:
Adjust default queue name to just be the prefix

As throttles are now per-prefix, any default queue strategy that
fragments messages sharing a common prefix amongst multiple pipes has an
effect where the despool logic will make very slow progress attempting
to send them.

This is particularly problematic for whispers. If an addon were to queue
up a lot of data for unique whisper targets, we can only despool one
message per second for at most one of those targets, with all the
other pipes constantly being blocked.

The worst case scenario is whisper data backs up faster than it can be
sent out on a prefix, and if the data itself is formed of multiple
messages then it may be the case that no user receives complete data in
a timely manner.

Because of this, the default queue name is now just the prefix itself.
This should ensure that in large transfers across a prefix at least one
player is still getting data in a more reasonable timeframe, and makes
delivery behavior more consistent with raw use of the API. This also
reduces the amount of work CTL has to do internally when despooling, as
by default there'll be less unique pipes.

Meorawr committed:
Add sendResult to AceComm callback forwarding

Most people probably interface with AceComm rather than CTL directly, so
it makes sense to pass the send result from CTL fully through the chain.

One downside is that because AceComm currently appends its own textLen
argument, we're forced to effectively sandwich it by appending
sendResult to the end of the list.

Meorawr commented:
Done some extremely light testing both on the PTR and Cata today; it appears to be fundamentally correct in terms of not bodging up transmission orders or losing messages.

One thing to note is that the ChannelThrottle status is a bit weird - waiting for some clarification on intended behaviors on this one.

For channel messages it seems you can send about 10 messages in a short period before the API returns the ChannelThrottle status for 5 seconds. The issue is that if you retry any sends on the channel in that 5 second window, it refreshes. Combine this with a periodic despool loop, and you can infinitely lock yourself out of channels.

Personally that feels like a bug in the API, and would much rather not have to duplicate the state that the client itself is maintaining re: channel throttles in the library when it could just simply _not_ refresh the timeout and allow CTL to work without special casing it with yet-another ring on a larger timer.

Nevcairiel commented:
Yes, the ChannelThrottle behavior sounds unfortunate and unintended. Tracking it and guessing on the timing sounds rather terrible, and I was glad that the status enum otherwise lets us just handle it without knowing or needing to know what the actual limits for any distribution actually are.

Meorawr commented:
So on ChannelThrottle there's interesting quirks.

Firstly, my prior test only considered looking at the return code from the point of the call and not monitoring what was actually received.

If ChannelThrottle is returned then there's no guarantees at all on any consistent behavior. It can queue it for transmission and ultimately deliver it with a significant delay on the order of 5-15 seconds, but it can also sometimes drop it. I suspect the actual behavior is there's a limited queue and it'll overwrite messages if you fill it up too fast before the delayed messages get sent off in a batch. Either way, it's impossible to tell whether or not the message will actually be sent with the return value in this case.

The good news however is this isn't new behavior. This is how it apparently has behaved for quite some time, so we probably don't need to think too much about it. My feeling is in the case of ChannelThrottle we should pessimistically treat is as a hard error - so don't requeue the message, and tell the caller that `didSend = false`. The message may arrive at its destination, but it should be assumed to be lost.

I'd also note that there's cases where the API can return Success and the message is dropped on CHANNELs anyway, so all of the above may also be the result of additional server-side throttles. Either way, it's a extremely lossy chat type.

Meorawr commented:
This is basically fully cooked at this point with the one exception of how we want to handle the blocked queue throttle. I'm open to either lowering it (0.5s) or flat out removing it for previously stated reasons.

More testing is also good, but so far I've not seen any issue of missing or unordered communications on both test realms nor live ones if I artifically limit messages on a prefix to 1 per second globally.

Separately - I've been rebasing #5 against these changes and also have another branch queued up that adds BNGameData support; it'd be good to get all three of these so that I can finally give Chomp some concrete shoes.

Meorawr commented on AceComm-3.0/ChatThrottleLib.lua:
```diff
@@ -292,38 +314,89 @@ end
+		local sendResult
 		local lowerDest = strlower(msg[3] or "")
+
 		if lowerDest == "raid" and not UnitInRaid("player") then
 			-- do nothing
+			sendResult = SendAddonMessageResult.NotInGroup
 		elseif lowerDest == "party" and not UnitInParty("player") then
 			-- do nothing
+			sendResult = SendAddonMessageResult.NotInGroup
```
Incidentally - do we want to delete these checks? I believe the API returns the NotInGroup result code in these cases anyway. Only issue is Classic Era which doesn't _yet_.

Nevcairiel replied:
If we're sure the chat warning message no longer comes up when attempting, sure.

Nevcairiel replied:
Tested on Retail, Cataclysm and Era, no longer seems to print errors into the chat window. As a side note- INSTANCE_CHAT always succeeds.

As an aside-aside, GUILD still prints "You are not in a guild.", but since we never cared to suppress those, might as well continue to not care.

Meorawr replied:
Removed these checks in the squashed version.

Nevcairiel commented:
I went over this again and it seems fine to me. Can you squash all the commits into its 3 logical changes for merging:
- Throttle handling
- AceComm change for the send result (separate library should have a separate commit)
- Change of the default pipe key to be the prefix

It should then be mostly ready, maybe pending a decision on one of the discussion points.

lantisnt commented on AceComm-3.0/ChatThrottleLib.lua:
```diff
@@ -292,38 +314,89 @@ end
+end
+
+local function IsThrottledSendResult(result)
+	return result == SendAddonMessageResult.AddonMessageThrottle
```
Any particular reason not to check for channelThrottle?

Meorawr replied:
Refer to the discussion in the comments: https://github.com/WoWUIDev/Ace3/pull/13#issuecomment-2091476663

ChannelThrottle is neither a new behavior, and that result code being returned still results in the messages sometimes being queued and delivered.

The key word is sometimes - there is an aspect of server-side throttling that's also enforced which may cause them to sometimes be dropped, but this isn't something that's particularly convenient for CTL to track because they _can_ still be submitted and delivered. If we attempt to retry, we may double-send (or worse, infinitely spam the message on repeat).

The only way CTL would be able to avoid double-sending is to completely implement its own throttled submission to channels from scratch without relying upon the API to return a suitable result.

Additionally there's a server-side bug that I was informed of that makes the throttling sometimes twice as aggressive as it should be. Either way, if you're hitting this return code then you should assume your messages have a <50% chance of ever making it through.

lantisnt replied:
So this boils down to wether its treated as drop and retry or assume delivered question? Should CTL try to guarantee delivery anyway by retrying it or inform higher level to try to retry? This will impact multipart ace3 communication anyway, but for single part the question is wether double delivery > no delivery.

Coming to think of it, adding ace3 unordered multipart support + ctl retries would be the best way to guarantee reliability.

lantisnt commented on AceComm-3.0/ChatThrottleLib.lua:
```diff
@@ -292,38 +314,89 @@ end
+				Prio.nTotalSent = Prio.nTotalSent + msg.nSize
+			end
+
+			-- Notify caller of delivery.
```
Comment is misleading as its not really delivery but a send

Meorawr replied:
Original comment, just minus a few words - but it makes sense to change it.

lantisnt commented on AceComm-3.0/ChatThrottleLib.lua:
```diff
@@ -353,40 +427,60 @@ function ChatThrottleLib.OnUpdate(this,delay)
-		if Prio.Ring.pos or Prio.avail < 0 then
-			n = n + 1
+	-- Integrate blocked queues back into their rings periodically.
+	if self.BlockedQueuesDelay >= 1 then
```
I do not see any gain in artificial delay added to readding blocked ones to ring. Putting queues at the end of `Ring` would be enough (as moving through ring should give you some delay) and would allow you to drop the `Blocked` ring completely.

One concern that I do not think I understand is what would happen if all queues are blocked at the same time?

Meorawr replied:
This was also discussed in the comments and is still pending resolution.

Meorawr commented:
One extra force push there to nuke the unused `lowerDest` variable, as that was only used by the now-dead checks for raid/party.
